File: src/django_bird/compiler.py
from __future__ import annotations

import re

BIRD_PATTERN = re.compile(
    r"<bird:(\w+)([^>]*)(?:/>|>(.*?)</bird:\1>)", re.DOTALL | re.MULTILINE
)


class Compiler:
    # placeholder for now, implementation to come
    def compile(self, input_string: str) -> str:
        tokens = self.tokenize(input_string)
        ast = self.parse(tokens)
        return self.transform(ast)

    def tokenize(self, input_string: str) -> str:
        return input_string

    def parse(self, tokens: str) -> str:
        return tokens

    def transform(self, ast: str) -> str:
        return ast

File: src/loader/src/compiler.rs
pub struct Compiler;

impl Compiler {
    pub fn compile<'a>(&self, input_string: &'a str) -> &'a str {
        let tokens = self.tokenize(input_string);
        let ast = self.parse(tokens);
        self.transform(ast)
    }

    fn tokenize<'a>(&self, input_string: &'a str) -> &'a str {
        input_string
    }

    fn parse<'a>(&self, tokens: &'a str) -> &'a str {
        tokens
    }

    fn transform<'a>(&self, ast: &'a str) -> &'a str {
        ast
    }
}
